Respect -UseWindowsPowerShell:$false in New-PSSession

yotsuda opened the pull request:
# PR Summary

This PR fixes the issue where `-UseWindowsPowerShell:$false` is not respected in `New-PSSession`, causing it to behave the same as `-UseWindowsPowerShell:$true`.

Fixes #26464
Related to #25242

## PR Context

When `-UseWindowsPowerShell:$false` is explicitly specified for `New-PSSession`, the cmdlet incorrectly creates a Windows PowerShell 5.1 session instead of a default PowerShell Core session. This occurs because the code checks the parameter set name (which is selected based on the *presence* of the switch, not its value) instead of checking the actual boolean value of the `UseWindowsPowerShell` parameter.

This fix changes the condition to directly check the `UseWindowsPowerShell` property value, ensuring that explicit `$false` values are properly respected.

### Changes

- Modified `newrunspacecommand.cs` line 464: Added `if (UseWindowsPowerShell)` check in the `UseWindowsPowerShellParameterSet` case
- Added test case in `PSSession.Tests.ps1` to verify `-UseWindowsPowerShell:$false` creates a default PowerShell Core session

### Testing

- Added new test: "Should create default PowerShell session when -UseWindowsPowerShell:$false is used"
- The test verifies that `-UseWindowsPowerShell:$false` creates a session with ShellUri matching the default PowerShell endpoint
- Pre-fix: Test fails (creates Windows PowerShell 5.1 session with ShellUri ending in "Microsoft.PowerShell.Workflow")
- Post-fix: Test passes (creates default session with ShellUri ending in "Microsoft.PowerShell")
- All existing tests continue to pass

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
  - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge. If this PR is a work in progress, please open this as a [Draft Pull Request and mark it as Ready to Review when it is ready to merge](https://docs.github.com/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests#draft-pull-requests).
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
  - [x] None
- **User-facing changes**
  - [x] Not Applicable
    - This fix corrects incorrect behavior to match expected behavior; no documentation changes needed
- **Testing - New and feature**
  - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

yotsuda committed:
Correct handling of explicit -UseWindowsPowerShell:$false parameter value in New-PSSession

iSazonov commented on test/powershell/engine/Remoting/PSSession.Tests.ps1:
```diff
@@ -91,3 +91,38 @@ Describe "SkipCACheck and SkipCNCheck PSSession options are required for New-PSS
+    It "Should respect explicit -UseWindowsPowerShell:`$false parameter value" {
+        # -UseWindowsPowerShell:$false should behave the same as not specifying the parameter
+        # When UseWindowsPowerShell is not specified (or explicitly false), it should attempt
+        # to connect via WinRM to localhost, which will fail if WinRM is not configured
```
We have tests for New-PSSession without exceptions.

yotsuda replied:
@iSazonov Thanks for the hint. I've updated the test to expect successful session creation instead of exceptions - verifying that :$true creates PowerShell 5.1 and :$false creates PowerShell Core via WinRM.

yotsuda committed:
Fix test to expect successful WinRM connection instead of exception

yotsuda commented on test/powershell/engine/Remoting/PSSession.Tests.ps1:
```diff
@@ -91,3 +91,53 @@ Describe "SkipCACheck and SkipCNCheck PSSession options are required for New-PSS
+            $session | Should -Not -BeNullOrEmpty
+
+            # Verify it's PowerShell Core (not Windows PowerShell 5.1)
+            $version = Invoke-Command -Session $session -ScriptBlock { $PSVersionTable.PSVersion }
```
Thank you for the feedback, @iSazonov.

I want to make sure I understand correctly. Should I update the test like this?
```powershell
# Test 2: -UseWindowsPowerShell:$false should create a PowerShell Core session via WinRM
$session = $null
try {
    $session = New-PSSession -UseWindowsPowerShell:$false -ConfigurationName PowerShell.7-preview -ErrorAction Stop
    $session | Should -Not -BeNullOrEmpty

    # Verify it's PowerShell Core (not Windows PowerShell 5.1)
    $version = Invoke-Command -Session $session -ScriptBlock { $PSVersionTable.PSVersion }
    $version.Major | Should -BeGreaterOrEqual 6
}
finally {
    if ($session) { Remove-PSSession $session -ErrorAction SilentlyContinue }
}
```

Please let me know if this is the right approach.

yotsuda committed:
Update test to verify -UseWindowsPowerShell:$false uses WSMan transport

- Adopt { ... } | Should -Not -Throw pattern as suggested by @mklement0
- Verify Transport is 'WSMan' (not 'Process') for -UseWindowsPowerShell:$false
- Simpler test that is independent of default PowerShell edition/version

yotsuda committed:
Address review feedback from iSazonov and Copilot

1. Rename $session1 to $sessionWithTrue for better symmetry with $sessionWithFalse
2. Add null check for $sessionWithFalse to match the pattern used in Test 1

These changes improve test code readability and robustness.
